client/my-sites/controller.js

```diff
@@ -210,44 +225,55 @@ module.exports = {
 			return next();
 		}
 
-		// If there's a valid site from the url path
-		// set site visibility to just that site on the picker
-		if ( sites.select( siteID ) ) {
-			const selectionComplete = onSelectedSiteAvailable( context );
+		// ensure we have fetched sites
+		const sitesFetchedPromise = new Promise( ( resolve, reject ) => {
+			if ( ! sites.fetched && sites.fetching ) {
+				sites.once( 'change', () => sites.fetched ? resolve() : reject() );
+			} else if ( currentUser.visible_site_count !== sites.getVisible().length ) {
+				sites.initialized = false;
+
+				const waitingNotice = notices.info( i18n.translate( 'Finishing set up…' ), { showDismiss: false } );
+				sites.once( 'change', () => {
+					notices.removeNotice( waitingNotice );
+					if ( sites.fetched ) {
+						sites.initialized = true;
+						return resolve();
+					}
 
-			// if there was a redirect, we should terminate processing of next routes
-			// and let the redirect proceed
-			if ( ! selectionComplete ) {
-				return;
-			}
-		} else {
-			// if sites has fresh data and siteID is invalid
-			// redirect to allSitesPath
-			if ( sites.fetched || ! sites.fetching ) {
-				return page.redirect( allSitesPath );
+					reject();
+				} );
+
+				sites.fetch();
+			} else {
+				return resolve();
 			}
+		} );
+
+		// Select a site
+		sitesFetchedPromise
+			.catch( error => {
+				page.redirect( allSitesPath );
+				return Promise.reject( error );
+			} )
+			.then( () => {
+				if ( ! sites.select( siteID ) ) {
+					return page.redirect( allSitesPath );
+				}
 
-			let waitingNotice;
-			const selectOnSitesChange = () => {
-				// if sites have loaded, but siteID is invalid, redirect to allSitesPath
-				if ( sites.select( siteID ) ) {
-					sites.initialized = true;
-					onSelectedSiteAvailable( context );
-					if ( waitingNotice ) {
-						notices.removeNotice( waitingNotice );
-					}
-				} else if ( ( currentUser.visible_site_count !== sites.getVisible().length ) ) {
-					sites.initialized = false;
-					waitingNotice = notices.info( i18n.translate( 'Finishing set up…' ), { showDismiss: false } );
-					sites.once( 'change', selectOnSitesChange );
```

Comment:
@yurynix I realized that another piece of behavior that isn't replicated by the change to using a promise here is the fact that sites are recursively fetched until the `visible_site_count` property of the user is equal to the list of visible sites from `sites-list`. I'm not sure if the refactoring approach with a Promise is going to be make it easy to replicate this. It might be a good idea to refactor this in a separate PR.